### ISSUE TITLE
Memory leak when running `npx tsx <filename>`, on a script that waits for the user input

### ISSUE THREAD

**Issue by felladrin** [esbuild-kit/tsx#229]
### Bug description

**Issue**: When running `npx tsx example.ts`, memory usage grows infinitely. It happens independently of the file extension (tested with `.ts`, `.js`, `.mjs`).

**Expectation**: Memory should be stable (~30 MB) while waiting for the user input.

**Confirmation**: For comparison, it doesn't happen when running `npx jiti example.ts`.

| tsx | jiti |
|--------|--------|
| <img width="825" alt="1" src="https://user-images.githubusercontent.com/418083/235093856-187084e0-7357-43ea-ac87-5a66f3bac461.png"> | <img width="826" alt="2" src="https://user-images.githubusercontent.com/418083/235093886-18765321-0b0e-4564-b7c5-efcf4a34df6e.png"> | 

### Reproduction

```typescript
// example.ts
import readline from "node:readline";

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.question("What is your name? ", (name) => {
  console.log(`Hello ${name}`);
  rl.close();
});
```

### Environment

```shell
System:
    OS: macOS 13.3.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 11.64 GB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.0.0 - /opt/homebrew/bin/node
    npm: 9.6.5 - /opt/homebrew/bin/npm
```

Note: I haven't installed `tsx` globally, but the version running through `npx` is the following:

```❯ npx tsx --version
tsx v3.12.6
node v20.0.0
```

### Can you work on a fix?

- [ ] I’m interested in opening a pull request to address this issue.

**Comment by privatenumber** [comment]
I'm unable to reproduce this and the reproduction doesn't feel sufficient.

Would like to see a better way to measure memory usage (e.g. via `process.memoryUsage()`). Currently, I'm not even sure if your Activity Monitor is reflecting the correct process.

(Also, shouldn't the baseline be Node.js?)

Closing this out as per the Issue guidelines for lack of reproduction.

**Comment by LangLangBart** [comment]
@felladrin [^1]
- can reproduce the issue since I updated to [nodejs/node](https://github.com/nodejs/node) **v20.0.0** (18/Apr/23) 
- switching back to **v19.9.0** the issue isn't reproducible 
  - tested with [Schniz/fnm](https://github.com/Schniz/fnm)

[^1]: Can you leave the name of your VS Code theme under your image ? Colors/font look great.

**Comment by felladrin** [comment]
Thanks for confirming, @LangLangBart![^1]

[^1]: The theme is [Stockholm](https://marketplace.visualstudio.com/items?itemName=LudvigLundgren.StockholmTheme) by @ludviglundgren 😄

**Comment by LangLangBart** [comment]
> The theme is Stockholm

thanks

---

- Tested the latest [nightly](https://github.com/nodejs/node#nightly-releases) node version on my intel mac 
- coudln't test the exact same setup as described  in the description as an error would occur
  - `All version numbers passed to esbuild must be in the format "X", "X.Y", or "X.Y.Z ... Invalid version: "21.0.0-nightly202304287ae1360358""`
- used a similar setup and the issue is no longer reproducible
  - I assume that this memory problem should be fixed with the next node release.
- Ref: https://github.com/nodejs/node/issues/47566

#### test setup
```sh
touch require.cjs
touch loader.mjs
cat << EOF > test.mjs
import { createInterface } from 'readline';

const rl = createInterface({
input: process.stdin,
output: process.stdout,
});

rl.question('Please enter your answer: ', () =>rl.close());
EOF
```

#### node v20
```sh
node --version
# v20.0.0

node --require ./require.cjs --loader ./loader.mjs test.mjs
# check memory % in a second terminal tab for node
while true; do ps -A -o %mem,comm | grep node; sleep 2; done
# 1.2 node
# 4.5 node
# 8.3 node
# 11.5 node
```
#### node latest nightly
```sh
curl --silent --remote-name https://nodejs.org/download/nightly/v21.0.0-nightly202304287ae1360358/node-v21.0.0-nightly202304287ae1360358-darwin-x64.tar.gz
tar -xzf node-v21.0.0-nightly202304287ae1360358-darwin-x64.tar.gz
export PATH="$HOME/node-v21.0.0-nightly202304287ae1360358-darwin-x64/bin:$PATH"

node --version
# v21.0.0-nightly202304287ae1360358

node --require ./require.cjs --loader ./loader.mjs test.mjs
while true; do ps -A -o %mem,comm | grep node; sleep 2; done
# 0.5 node
# 0.5 node
# 0.5 node
# 0.5 node
```

**Comment by privatenumber** [comment]
Thanks for the thorough investigation @LangLangBart

Glad it's not an issue with tsx. But we should fix the nightly error:
> All version numbers passed to esbuild must be in the format "X", "X.Y", or "X.Y.Z ... Invalid version: "21.0.0-nightly202304287ae1360358""

**Comment by LangLangBart** [comment]
> Thanks for the thorough investigation @LangLangBart

:grinning:

---

@privatenumber

- related issue in `evanw/esbuild` about node nightly versions
  - https://github.com/evanw/esbuild/issues/2766
- seems to be intentional

**Comment by felladrin** [comment]
Hey there! I just wanted to let you know that this issue has been resolved in [Node v20.1.0](https://nodejs.org/en/blog/release/v20.1.0)  🎉
